src/bare-metal/microcontrollers/examples/src/bin/mmio.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// ANCHOR: Example
+#![no_main]
+#![no_std]
+
+extern crate panic_halt as _;
+
+use core::mem::size_of;
+use cortex_m_rt::entry;
+
+/// GPIO port 0 peripheral address
+const GPIO_P0: usize = 0x5000_0000;
+
+// GPIO peripheral offsets
+const PIN_CNF: usize = 0x700;
+const OUTSET: usize = 0x508;
+const OUTCLR: usize = 0x50c;
+
+// PIN_CNF fields
+const DIR_OUTPUT: u32 = 0x1;
+const INPUT_DISCONNECT: u32 = 0x1 << 1;
+const PULL_DISABLED: u32 = 0x0 << 2;
```

Comment:
Ehm, this is just zero, right? Why would one shift a zero bit like this?

Reply:
Bit 2 specifies whether the pull-up resistor is disabled (0) or enabled (1). It's good practice to be explicit about each setting we are configuring. Usually I'd write out the full set of constants for all the enum values, but to save space here I've only included the ones I actually use.